src/System.Net.Security/src/System/Net/Security/SNIHelper.cs

```diff
@@ -0,0 +1,251 @@
+        // ClientHello structure:
+        //   - ProtocolVersion client_version (2 bytes)
+        //   - Random random (32 bytes => 4 bytes GMT unix timestamp + 28 bytes of random bytes)
+        //   - SessionID session_id (opaque type of max size 32 => size fits in 1 byte)
+        //   - CipherSuite cipher_suites (opaque type of max size 2^16-1 => size fits in 2 bytes)
+        //   - CompressionMethod compression_methods (opaque type of max size 2^8-1 => size fits in 1 byte)
+        //   - Extension client_hello_extension_list (opaque type of max size 2^16-1 => size fits in 2 bytes)
+        private static string GetSniFromClientHello(ReadOnlySpan<byte> clientHello)
+        {
+            // Skip ProtocolVersion and Random
+            ReadOnlySpan<byte> p = SkipBytes(clientHello, 34);
+
+            // Skip SessionID
+            p = SkipOpaqueType1(p);
+
+            // Skip cipher suites
+            p = SkipOpaqueType2(p);
+
+            // Skip compression methods
+            p = SkipOpaqueType1(p);
+
+            // is invalid structure or no extensions?
+            if (p.IsEmpty)
+            {
+                return null;
+            }
+
+            ushort extensionListLength = ReadUint16(p);
+            p = SkipBytes(p, 2);
+
+            if (extensionListLength != p.Length)
+            {
+                return null;
+            }
+
+            while (!p.IsEmpty)
+            {
+                string sni = GetSniFromExtension(p, out p);
+                if (sni != null)
+                {
+                    return sni;
+                }
+            }
+
+            return null;
+        }
+
+        // 2.3. https://www.ietf.org/rfc/rfc3546.txt
+        // Extension structure:
+        //   - ExtensionType extension_type (2 bytes) => 0x00 is server_name
+        //   - opaque extension_data
+        private static string GetSniFromExtension(ReadOnlySpan<byte> extension, out ReadOnlySpan<byte> remainingBytes)
+        {
+            if (extension.Length < 2)
+            {
+                remainingBytes = ReadOnlySpan<byte>.Empty;
+                return null;
+            }
+
+            ushort extensionType = ReadUint16(extension);
+            ReadOnlySpan<byte> extensionData = extension.Slice(2);
+
+            if (extensionType == 0x00)
+            {
+                return GetSniFromServerNameList(extensionData, out remainingBytes);
+            }
+            else
+            {
+                remainingBytes = SkipOpaqueType2(extensionData);
+                return null;
+            }
+        }
+
+        // 3.1. https://www.ietf.org/rfc/rfc3546.txt
+        // ServerNameList structure:
+        //   - ServerName server_name_list<1..2^16-1>
+        // ServerName structure:
+        //   - NameType name_type (1 byte) => 0x00 is host_name
+        //   - opaque HostName
+        // Per spec:
+        //   If the hostname labels contain only US-ASCII characters, then the
```

Comment:
this part is TODO. We will likely decide on either returning bytes or IDN unmap but let's wait for API design discussion - for now treating as UTF-8 string